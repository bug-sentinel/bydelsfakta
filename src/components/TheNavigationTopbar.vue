<template>
  <header class="oslo__navigation-topbar" @keydown.esc="showDropdown = false">
    <h1 v-if="selectedSubpage === null" class="header">{{ getDistrict(this.$route.params.district) }}</h1>
    <div class="navigation-topbar" :class="{ 'navigation-topbar--hidden': selectedSubpage === null }">
      <button
        v-if="selectedSubpage"
        id="select"
        class="navigation-topbar__select"
        @click="showDropdown = !showDropdown"
        @keydown.enter="showDropdown = !showDropdown"
        @keydown.escape="closeMenu"
        v-click-outside="closeMenu"
        ref="select"
        :aria-label="$t('navigationTopbar.selectTopic.aria')"
      >
        <label class="label" :class="{ 'label--active': selectedSubpage !== null }">{{
          $t('navigationTopbar.selectTopic.label')
        }}</label>
        <div class="topic">
          <span>{{ getHumanReadableTopic(selectedSubpage) }}</span>
          <i class="material-icons">{{ showDropdown ? 'arrow_drop_up' : 'arrow_drop_down' }}</i>
        </div>
      </button>
      <transition name="fade">
        <div id="dropdown" class="navigation-topbar__dropdown" v-if="showDropdown">
          <div v-for="(kategori, index) in dropdown" :key="index" class="navigation-topbar__dropdown-column">
            <div
              class="navigation-topbar__dropdown-column--heading"
              :style="{ color: kategori.color, 'border-top': `3px solid ${kategori.color}` }"
            >
              <span>{{ kategori.kategori }}</span>
            </div>
            <div class="navigation-topbar__dropdown-items">
              <router-link
                v-for="(link, subpageIndex) in kategori.links"
                class="navigation-topbar__dropdown-item"
                :id="`dropdown-href-${topics[link].value}`"
                :class="{
                  'navigation-topbar__dropdown-item--active': checkActiveSubpage(topics[link].value),
                  'navigation-topbar__dropdown-item--disabled': disabledTopics.find(topic => topic === topics[link].value),
                }"
                :key="subpageIndex"
                v-text="topics[link].text"
                :to="onClickSubpage(topics[link].value)"
                >{{ topics[`${link}`].text }}</router-link
              >
            </div>
          </div>
        </div>
      </transition>
    </div>
  </header>
</template>
<script>
import { mapState, mapActions } from 'vuex';
import allDistricts from '../config/allDistricts';
import { categories, topics, disabledTopics } from '../config/topics';

export default {
  name: 'TheNavigationTopbar',
  data() {
    return {
      selectedSubpage: null,
      dropdown: categories,
      allDistricts: allDistricts,
      showDropdown: false,
      topics: topics,
      disabledTopics: disabledTopics,
    };
  },

  computed: {
    ...mapState(['compareDistricts', 'menuIsOpen']),
  },

  created() {
    if (this.$route.name !== 'Home') {
      const routes = this.$route.path.split('/');
      if (routes.length > 3) this.selectedSubpage = routes[3];
    }
  },

  methods: {
    ...mapActions(['setMenuIsOpen']),

    closeMenu() {
      if (this.showDropdown) {
        this.showDropdown = false;
      }
    },

    getHumanReadableTopic(id) {
      if (this.topics[id]) {
        return this.topics[id].text;
      } else {
        this.selectedSubpage = null;
        return '';
      }
    },

    getDistrict(id) {
      if (this.compareDistricts || id === 'alle') {
        return this.$t('navigationTopbar.header.compareDistrict');
      } else if (this.$route.name === 'NotFound') {
        return this.$t('navigationTopbar.header.notFound');
      } else {
        return id !== undefined
          ? this.allDistricts.find(district => district.uri === id).value
          : this.$t('navigationTopbar.header.chooseDistrict');
      }
    },

    checkActiveSubpage(subpage) {
      return this.$route.path.includes(subpage);
    },

    onClickSubpage(subpage) {
      return this.disabledTopics.find(topic => topic === subpage)
        ? ''
        : { name: 'Topic', params: { district: this.$route.params.district, topic: subpage } };
    },
  },

  watch: {
    $route(to) {
      const routes = to.path.split('/');
      if (to.name === 'NotFound') {
        this.selectedSubpage = null;
        this.showDropdown = false;
      } else if (to.name === 'District') {
        this.selectedSubpage = null;
      } else if (to.name === 'Home') {
        this.selectedSubpage = null;
      } else {
        if (routes.length > 3) this.selectedSubpage = routes[3];
      }
    },

    showDropdown() {
      this.setMenuIsOpen(this.showDropdown);
    },
  },
};
</script>
<style scoped lang="scss">
@import '../styles/colors';
@import '../styles/variables';

.header {
  color: $color-purple;
  flex-direction: column;
  font-size: $font-huge;
  position: relative;
  width: 100%;
  font-weight: 500;
  padding: 0.5rem;
  margin-left: 0.5rem;
  display: none;

  @media screen and (min-width: $break-lg) {
    display: block;
  }
}

.oslo__navigation-topbar {
  display: flex;
  flex-direction: column;
  margin: 1.5rem 0.5rem 0.5rem 0.5rem;
  width: calc(100% - 1rem);
  position: relative;
  top: 0;
  background-color: $color-bg;
  z-index: 3;
  padding: 1rem 0;

  @media screen and (min-width: $break-lg) {
    padding: 2.5rem 1rem 1rem;
    width: 100%;
    margin: 0;
  }

  &-button {
    display: none;
    flex-direction: row;

    @media screen and (min-width: $break-lg) {
      display: flex;
    }
  }
}

.oslo__topbar {
  color: $color-purple;
  font-weight: bold;
  margin-bottom: 2rem;

  &-text {
    letter-spacing: 0.8px;
    margin-left: 1rem;
    margin-top: 0.3rem;
    text-transform: uppercase;
  }
}

.label {
  font-size: $font-body;
  margin-bottom: 0.25rem;
}

.topic {
  display: flex;
  align-items: center;
}

.navigation-topbar {
  width: 100%;
  max-width: 768px;
  position: relative;

  &--hidden {
    @media screen and (max-width: $break-lg) {
      display: none;
    }
  }

  &__select {
    color: $color-purple;
    display: flex;
    flex-direction: column;
    font-size: $font-huge;
    position: relative;
    width: 100%;
    font-weight: 500;
    padding: 0.5rem;

    @media screen and (min-width: $break-md) {
      padding: 1rem;
    }

    &:focus {
      outline: none;
    }

    &:hover {
      background-color: darken($color-bg, 5%);
    }
  }

  &__dropdown {
    background-color: white;
    border: 1px solid $color-grey-100;
    box-shadow: 0 2px 3px 0 rgba(0, 0, 0, 0.65);
    border-bottom-left-radius: 2px;
    border-bottom-right-radius: 2px;
    width: 100%;
    max-width: 768px;
    display: flex;
    flex-flow: row wrap;
    position: absolute;
    z-index: 1;
    padding: 1rem 0.5rem;

    max-height: calc(100vh - 17rem);
    overflow: auto;

    &-column {
      display: flex;
      flex-direction: column;
      flex: 33.333%;
      min-width: 200px;
      padding: 0.5rem;

      &--heading {
        font-weight: bold;
        margin-bottom: 1rem;
        margin-top: 1rem;

        @media screen and (min-width: $break-lg) {
          margin-top: 0;
        }
      }
    }

    &-items {
      display: flex;
      flex-direction: column;
    }

    &-item {
      background-color: $color-grey-50;
      border-radius: 4px;
      color: $color-purple;
      margin-bottom: 0.5rem;
      padding: 1rem 0.75rem;

      @media screen and (min-width: $break-lg) {
        margin-bottom: 0.1rem;
        padding: 0.5rem;
      }

      &:hover:not(&--active):not(&--disabled) {
        background-color: $color-grey-100;
        color: black;
      }

      &--active {
        background-color: $color-blue;
      }

      &--disabled {
        opacity: 0.45;
        background-color: white;
      }
    }
  }
}
</style>
