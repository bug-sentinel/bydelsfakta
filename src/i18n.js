import Vue from 'vue';
import VueI18n from 'vue-i18n';

function loadLocaleMessages() {
  const locales = require('./locales/nb-no.json');
  return {
    'nb-no': locales,
  };
}

export default function setupI18n() {
  Vue.use(VueI18n);

  return new VueI18n({
    locale: process.env.VUE_APP_I18N_LOCALE || 'nb-no',
    fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'nb-no',
    messages: loadLocaleMessages(),
  });
}
