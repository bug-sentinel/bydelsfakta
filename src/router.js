import Vue from 'vue';
import Router from 'vue-router';
import Home from './views/Home.vue';
import Bydel from './views/Bydel.vue';
import Alder from './views/subviews/Alder.vue';
import Boligpriser from './views/subviews/Boligpriser.vue';
import Bygningstyper from './views/subviews/Bygningstyper.vue';
import Eierform from './views/subviews/Eierform.vue';
import Flytting from './views/subviews/Flytting.vue';
import Folkemengde from './views/subviews/Folkemengde.vue';
import Husholdninger from './views/subviews/Husholdninger.vue';
import Innvandrerbefolkningen from './views/subviews/Innvandrerbefolkningen.vue';
import Kommunaleboliger from './views/subviews/Kommunaleboliger.vue';
import Levekaar from './views/subviews/Levekaar.vue';
import Trangboddhet from './views/subviews/Trangboddhet.vue';

import Graph from './components/Graph.vue';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'Home',
      component: Home,
    },
    {
      path: '/graphA',
      name: 'Graph1',
      props: { settings: { url: '/mockDataA.json', template: 'a' } },
      component: Graph,
    },
    {
      path: '/graphB',
      name: 'Graph2',
      props: { settings: { url: '/mockDataA2.json', template: 'a' } },
      component: Graph,
    },
    {
      path: '/graphC',
      name: 'Graph3',
      props: { settings: { url: '/mockDataB.json', template: 'b' } },
      component: Graph,
    },
    {
      path: '/bydel/:bydel',
      name: 'Bydel',
      component: Bydel,
    },
    {
      path: '/bydel/:bydel/folkemengde',
      name: 'BydelFolkemengde',
      component: Folkemengde,
    },
    {
      path: '/bydel/:bydel/alder',
      name: 'BydelAlder',
      component: Alder,
    },
    {
      path: '/bydel/:bydel/innvandrerbefolkningen',
      name: 'BydelInnvandrerbefolkningen',
      component: Innvandrerbefolkningen,
    },
    {
      path: '/bydel/:bydel/husholdninger',
      name: 'BydelHusholdninger',
      component: Husholdninger,
    },
    {
      path: '/bydel/:bydel/trangboddhet',
      name: 'BydelTrandboddhet',
      component: Trangboddhet,
    },
    {
      path: '/bydel/:bydel/eierform',
      name: 'BydelEierform',
      component: Eierform,
    },
    {
      path: '/bydel/:bydel/kommunaleboliger',
      name: 'BydelKommunaleboliger',
      component: Kommunaleboliger,
    },
    {
      path: '/bydel/:bydel/flytting',
      name: 'BydelFlytting',
      component: Flytting,
    },
    {
      path: '/bydel/:bydel/bygningstyper',
      name: 'BydelBygningstyper',
      component: Bygningstyper,
    },
    {
      path: '/bydel/:bydel/boligpriser',
      name: 'BydelBoligpriser',
      component: Boligpriser,
    },
    {
      path: '/bydel/:bydel/levekaar',
      name: 'Bydellevekaar',
      component: Levekaar,
    },
    {
      path: '/sammenlign/:bydel*',
      name: 'Sammenlign',
      component: Bydel,
      children: [
        {
          path: '/folkemengde',
          name: 'Folkemengde',
          component: Folkemengde,
        },
        {
          path: '/alder',
          name: 'Alder',
          component: Alder,
        },
        {
          path: '/innvandrerbefolkningen',
          name: 'Innvandrerbefolkningen',
          component: Innvandrerbefolkningen,
        },
        {
          path: '/husholdninger',
          name: 'Husholdninger',
          component: Husholdninger,
        },
        {
          path: '/trangboddhet',
          name: 'Trandboddhet',
          component: Trangboddhet,
        },
        {
          path: '/eierform',
          name: 'Eierform',
          component: Eierform,
        },
        {
          path: '/kommunaleboliger',
          name: 'Kommunaleboliger',
          component: Kommunaleboliger,
        },
        {
          path: '/flytting',
          name: 'Flytting',
          component: Flytting,
        },
        {
          path: '/bygningstyper',
          name: 'Bygningstyper',
          component: Bygningstyper,
        },
        {
          path: '/boligpriser',
          name: 'Boligpriser',
          component: Boligpriser,
        },
        {
          path: '/levekaar',
          name: 'levekaar',
          component: Levekaar,
        },
      ],
    },
  ],
});
