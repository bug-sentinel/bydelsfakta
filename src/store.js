import Vue from 'vue';
import Vuex from 'vuex';
import router from './router';

import districts from './config/geoData/districts';
import bydeler from './config/allDistricts';

Vue.use(Vuex);

const state = {
  compareDistricts: false,
  districts: [],
  districtsGeo: districts,
};

const getters = {
  geoDistricts: state => {
    if (!state.compareDistricts && state.districts.length !== 0) {
      return { ...state.districtsGeo[`${bydeler.find(district => district.key === state.districts[0]).uri}`] };
    } else if (state.districts[0] === 'alle') {
      return { ...state.districtsGeo.oslo };
    }

    const features = state.districts.map(id =>
      state.districtsGeo.oslo.features.find(district => district.properties.id === parseInt(id))
    );

    return {
      ...state.districtsGeo.oslo,
      features,
    };
  },
};

const mutations = {
  ADD_DISTRICT(state, payload) {
    state.compareDistricts = true;
    state.districts = payload;
  },
  SELECT_DISTRICT(state, payload) {
    state.compareDistricts = false;
    state.districts = payload;
  },
  CLEAN_STATE(state) {
    state.compareDistricts = false;
    state.districts = [];
    state.districtsGeo = districts;
  },
};

const actions = {
  addDistrict({ commit }, payload) {
    const districts = payload.district.split('-');
    if (districts.length === 1) {
      if (districts[0] === 'alle') {
        commit('ADD_DISTRICT', districts);
      } else {
        const districtValue = bydeler.find(district => district.uri === districts[0]);
        const districtKey = bydeler.find(district => district.key === districts[0]);
        districtValue === undefined
          ? commit('ADD_DISTRICT', [districtKey.key])
          : commit('SELECT_DISTRICT', [districtValue.key]);
      }
    } else {
      commit('ADD_DISTRICT', districts);
    }

    if (payload.pushRoute) {
      router.currentRoute.params.topic === undefined
        ? router.push({ name: 'District', params: { district: districts.join('-') } })
        : router.push({
            name: 'Topic',
            params: { district: districts.join('-'), topic: router.currentRoute.params.topic },
          });
    }
  },
  cleanState({ commit }) {
    commit('CLEAN_STATE');
  },
};

export default new Vuex.Store({
  strict: true,
  state,
  getters,
  mutations,
  actions,
});
