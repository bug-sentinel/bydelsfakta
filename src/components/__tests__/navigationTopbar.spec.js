import { mount } from '@vue/test-utils';
import router from '@/router';
import clickOutside from '@/directives/clickOutside';
import store from '@/store';
import i18n from '@/i18n';
import TheNavigationTopbar from '../TheNavigationTopbar.vue';

global.scroll = jest.fn();
window.scroll = jest.fn();

describe('TheNavigationTopbar', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = mount(TheNavigationTopbar, {
      global: {
        plugins: [router, store, i18n],
        directives: {
          'click-outside': clickOutside,
        },
        stubs: {
          'ok-icon': true,
        },
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders navigationDrawer-component and finds oslo__navigation-topbar-class', () => {
    expect(wrapper.classes('oslo__navigation-topbar')).toBe(true);
  });

  test('renders correctly', async () => {
    await router.push('/bydel/sagene');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('return false if subpage is not active', async () => {
    await router.push('/bydel/sagene/boligpriser');
    expect(wrapper.vm.checkActiveTopic('levekaar')).toEqual(false);
  });

  test('change showDropdown to false if it is true', async () => {
    wrapper.setData({ showDropdown: true });
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('keep showDropdown as false if false', async () => {
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('return router object when clicking on a subpage', async () => {
    await router.push('/bydel/sagene/boligpriser');
    expect(wrapper.vm.onClickTopic('boligpriser')).toEqual({
      name: 'Topic',
      params: { district: 'sagene', topic: 'boligpriser' },
    });
  });
});
