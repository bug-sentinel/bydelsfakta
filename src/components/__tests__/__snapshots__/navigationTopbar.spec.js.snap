// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheNavigationTopbar renders correctly 1`] = `
<header
  class="oslo__navigation-topbar"
  id="main"
>
  <h1
    class="header"
  >
    Bydelsfakta – statistikk om befolkning, levekår og boforhold
  </h1>
   
  <div
    class="navigation-topbar navigation-topbar--hidden"
  >
    <!---->
     
    <!---->
  </div>
</header>
`;
